controllers/clustertopology_controller_compute.go

```diff
@@ -69,10 +73,116 @@ type machineDeploymentTopologyState struct {
-	// TODO: add get class logic; also remove nolint exception from clusterTopologyClass and machineDeploymentTopologyClass
-	return nil, nil
+// getClass gets the ClusterClass and the referenced templates to be used for a managed Cluster topology. It also converts
+// and patches all ObjectReferences in ClusterClass and ControlPlane to the latest apiVersion of the current contract.
+// NOTE: This function assumes that cluster.Spec.Topology.Class is set.
+func (r *ClusterTopologyReconciler) getClass(ctx context.Context, cluster *clusterv1.Cluster) (_ *clusterTopologyClass, reterr error) {
+	class := &clusterTopologyClass{
+		clusterClass:       &clusterv1.ClusterClass{},
+		machineDeployments: map[string]machineDeploymentTopologyClass{},
+	}
+
+	// get ClusterClass
+	if err := r.Client.Get(ctx, client.ObjectKey{Name: cluster.Spec.Topology.Class, Namespace: cluster.Namespace}, class.clusterClass); err != nil {
+		return nil, errors.Wrapf(err, "failed to retrieve ClusterClass %q in namespace %q", cluster.Spec.Topology.Class, cluster.Namespace)
+	}
+
+	// We use the patchHelper to patch potential changes to the ObjectReferences in ClusterClass.
+	patchHelper, err := patch.NewHelper(class.clusterClass, r.Client)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err := patchHelper.Patch(ctx, class.clusterClass); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, errors.Wrapf(err, "failed to patch ClusterClass %q in namespace %q", class.clusterClass.Name, class.clusterClass.Namespace)})
+		}
+	}()
+
+	// get ClusterClass.spec.infrastructure
+	if class.infrastructureClusterTemplate, err = r.convertReferenceAPIContractAndGet(ctx, class.clusterClass.Spec.Infrastructure.Ref); err != nil {
+		return nil, err
+	}
+
+	// get ClusterClass.spec.controlPlane
+	if class.controlPlane.template, err = r.convertReferenceAPIContractAndGet(ctx, class.clusterClass.Spec.ControlPlane.Ref); err != nil {
+		return nil, err
+	}
+
+	if class.clusterClass.Spec.ControlPlane.MachineInfrastructure != nil && class.clusterClass.Spec.ControlPlane.MachineInfrastructure.Ref != nil {
+		// get ClusterClass.spec.controlPlane.machineInfrastructure
+		if class.controlPlane.infrastructureMachineTemplate, err = r.convertReferenceAPIContractAndGet(ctx, class.clusterClass.Spec.ControlPlane.MachineInfrastructure.Ref); err != nil {
+			return nil, err
+		}
+	}
+
+	// if ControlPlaneTemplate.spec.template.spec.machineTemplate.infrastructureRef is set, convert and patch it
+	// returns an error if the ObjectReference is partially set
+	_, ok, err := unstructured.NestedMap(class.controlPlane.template.Object, "spec", "template", "spec", "machineTemplate", "infrastructureRef")
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to check existence of spec.template.spec.machineTemplate.infrastructureRef in %s %q in namespace %q", class.controlPlane.template.GetKind(), class.controlPlane.template.GetName(), class.controlPlane.template.GetNamespace())
+	}
```

Comment:
Should this be a method on `class`?

Reply:
Sounds reasonable (in case we keep it; https://github.com/kubernetes-sigs/cluster-api/pull/5007#discussion_r675452485)

Reply:
Dropped the code